### PR TITLE
Speed up pauli evolve clifford

### LINKED ISSUE #8177: pauli.evolve() needs a major speedup
### What should we add?

@aeddins-ibm found that `pauli.evolve()` can be quite slow, and we can get substantial performance improvement from using private methods for special cases. 

Since `pauli.evolve()` is the main entry-point / user-interface, it should be fast. There are probably some simple things that can be done to dramatically speed it up. But it needs some profiling to see what's causing the slowdown.

Here's a code that shows `pauli.evolve()` is 10,000x slower than calling `_evolve_cx()` on individual instructions. Some of this is expected since CX is a special case of clifford. But a) we should have faster ways of dealing with special cases of self-adjoint cliffords and b) is there some other overhead from validation/type-checking?

```py
from qiskit import QuantumCircuit
from qiskit.quantum_info.operators.symplectic.base_pauli import _evolve_cx
from qiskit.quantum_info import Clifford, Pauli, PauliList, random_pauli_list, SparsePauliOp
import time

n = 65
paulis_per_instance = 100
instances = 100

qc = QuantumCircuit(n)
qc.cx(range(n-1),range(1,n))
cliff = Clifford(qc)

result1 = random_pauli_list(n, paulis_per_instance*instances)
result2 = result1.copy()

start = time.time()
result1 = result1.evolve(cliff,frame='s')
print('time using evolve:',time.time()-start)

start = time.time()
for gate in qc.data:
    _evolve_cx(result2, gate[1][0]._index, gate[1][1]._index)
print('time using _evolve_cx:',time.time()-start)

print('same answer?',result1 == result2)
```

```py
time using evolve: 33.13418793678284
time using _evolve_cx: 0.0023658275604248047
same answer? True
```

### PR DESCRIPTION
**Pull request by aeddins-ibm**
Per issue #8177 . Seems to give a ~100x speed up by vectorizing for loops on small test cases.

I am not sure about the memory cost of the vectorization (einsum, etc).

For small-ish test cases, the for-loop evolve is faster than this version in the limit of many qubits and shorter PauliList length. For a single Pauli, the break-even on my laptop is about 150 qubits. The break even goes up for longer PauliLists, at least memory permitting (for length 4 list, break even on my laptop is ~250 qubits).

First time making a PR, sorry if there's more needed re: tests, styling, etc.

fixes #8177

### CODE REVIEW

**Comment by mtreinish** [comment]
There are some code formatting issues that `black` is reporting. You can just run black locally (typically with `tox -eblack`) to have these automatically fixed. More details on the formatting and lint rules are here: https://github.com/Qiskit/qiskit-terra/blob/main/CONTRIBUTING.md#style-and-lint

**Comment by aeddins-ibm** [comment]
Here is a quick test using time.time(), calling evolve 100 times with random pauli and random clifford arguments for several cases, and comparing the distribution of times taken using the old nested-for code (blue) and the new code (orange). The old and new code results agreed for all cases. The bottom row shows that the major slowdown motivating issue #8177 is significantly improved, so while room for improvement remains I think this PR can close that issue.

image

**Review by ShellyGarion — APPROVED** [review]
LGTM. Since there are no API changes, just making the code much faster, I would be happy if it will be merged before the coming release.

File: qiskit/quantum_info/operators/symplectic/base_pauli.py [hunk 1]
```diff
@@ -279,43 +279,39 @@ def evolve(self, other, qargs=None, frame="h"):
...
-        ret.dot(PauliList(pauli_list), qargs=qargs, inplace=True)
-
-        pauli_list = []
-        for x in self._x[:, idx]:
-            pauli = Pauli("I" * num_act)
-            for row in adj.destabilizer[x]:
-                pauli.compose(Pauli((row.Z[0], row.X[0], 2 * row.phase[0])), inplace=True)
-            pauli_list.append(pauli)
-        ret.dot(PauliList(pauli_list), qargs=qargs, inplace=True)
+        is_pauli_list = isinstance(self, PauliList)
```
Comment (jakelishman) [inline qiskit/quantum_info/operators/symplectic/base_pauli.py:298]: We shouldn't be `isinstance` checking for subclasses in a base class - it produces a two-way coupling in the inheritance tree that fragments subclass logic.  If `PauliList` has a valid reason to define that it evolves differently under Clifford operations, we should ensure that `_evolve_clifford` is safe for subclasses to override, and have the extra logic in `PauliList`.

That doesn't necessarily mean that all the set-up logic has to be duplicated.  In this situation, it's probably easiest to do that, but an alternative is to have `BasePauli` do the set-up and input normalisation, then call an abstract private method that subclasses _must_ override to do the "core" logic (how that's defined depends on the problem), and then the base-class method does the tidy-up / output formatting at the end.  It's a bit of a judgment call if there's a natural split here that makes that appropriate, or if it's better not to have the split, and just have a little bit of code duplication.

There's an added complication here that `BasePauli` is the base class of `Pauli`, but that appears (on a quick glance) to be a pretty artificial separation to me, and one whose abstraction leaks quite a lot.  That's of course pre-existing, and being clearly about how base-class/subclass relations (e.g. overridden methods) is how we move towards fixing that.

**Review by jakelishman — COMMENTED** [review]
I'm sorry, the code changes are too non-trivial to waive the missed release deadline on the 9th (that's nobody's fault, this issue just didn't get discussed until the deadline was already on us), even if it doesn't change the public API.  We can merge it as soon as 0.21 is tagged, though.

**Comment by jakelishman** [comment]
Just to be clear, I'm happy to go ahead with merging the code in its current state after the frozen release goes out, and I'm super pleased with the speed-ups, thanks!

When we have big performance improvements, we often sing about them in "feature" release notes - you can add one if you like.  You do `reno new --edit improve-pauli-evolve-clifford` or something, delete everything but the "feature" section, and write a sentence about the speed up.

**Comment by jakelishman** [comment]
Oh no, sorry.  Thanks for pointing it out.  This one is just an oversubscribed virtual machine on Azure that caused a timeout - I'll trigger a rerun and the PR should just continue to merge.  No need for you to do anything - the bot will update the branch at an appropriate time.

**Pull request merged**